api/sys_raft.go

```diff
@@ -0,0 +1,82 @@
+package api
+
+import (
+	"context"
+	"io"
+)
+
+// RaftJoinResponse represents the response of the raft join API
+type RaftJoinResponse struct {
+	Joined bool `json:"joined"`
+}
+
+// RaftJoinRequest represents the parameters consumed by the raft join API
+type RaftJoinRequest struct {
+	LeaderAddr string `json:"leader_api_addr"`
+	CACert     string `json:"ca_cert":`
+	Retry      bool   `json:"retry"`
+}
+
+// RaftJoin adds the node from which this call is invoked from to the raft
+// cluster represented by the leader address in the parameter.
+func (c *Sys) RaftJoin(opts *RaftJoinRequest) (*RaftJoinResponse, error) {
+	r := c.c.NewRequest("POST", "/v1/sys/storage/raft/join")
+
+	if err := r.SetJSONBody(opts); err != nil {
+		return nil, err
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var result RaftJoinResponse
+	err = resp.DecodeJSON(&result)
+	return &result, err
+}
+
+// RaftSnapshot invokes the API that takes the snapshot of the raft cluster and
+// writes it to the supplied io.Writer.
+func (c *Sys) RaftSnapshot(snapWriter io.Writer) error {
+	r := c.c.NewRequest("GET", "/v1/sys/storage/raft/snapshot")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(snapWriter, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RaftSnapshotRestore reads the snapshot from the io.Reader and installs that
+// snapshot, returning the cluster to the state defined by it.
+func (c *Sys) RaftSnapshotRestore(snapReader io.Reader, force bool) error {
```

Comment:
Should this be `snapRead io.ReadCloser`?

Reply:
I didn't think so. The reasoning I had was that the caller of this API should take care of closing it. This is what the command `vault operator raft restore` is doing. It open the file for reading, defers its close and passes the reader down to the API. That way API isn't taking the responsibility of closing something that is not supposed to. Does this sound like a proper workflow? Let me know if you think otherwise.

Reply:
Ah right. I was looking downstream and eventually found `retryablehttp.Client.Do` calling `resp.Body.Close()`, but it wraps the reader around `ioutil.NopCloser` which is why it can be done, but yeah the caller should be closing in this case.